src/erlcloud_aws.erl

```diff
@@ -574,6 +590,51 @@ prop_to_list_defined( Name, Props ) ->
     end.
 
 
+-spec get_role_credentials(aws_config()) -> {ok, #role_credentials{}} | {error, term()}.
+get_role_credentials(#aws_config{assume_role = AssumeRoleOptions} = Config) ->
+    RoleArn = proplists:get_value(role_arn, AssumeRoleOptions, undefined),
+    case application:get_env(erlcloud, {role_credentials, RoleArn}) of
+        {ok, #role_credentials{expiration_gregorian_seconds = Expiration} = Credentials} ->
+            Now = calendar:datetime_to_gregorian_seconds(calendar:universal_time()),
+            %% Get new credentials if these will expire in less than 2 minutes
+            case Expiration - Now < 120 of
+                true -> get_credentials_from_role(Config);
+                false -> {ok, Credentials}
+            end;
+        undefined ->
+            get_credentials_from_role(Config)
+    end.
+
+-spec get_credentials_from_role(aws_config()) -> {ok, #role_credentials{}} |
+                                                 {error, term()}.
+get_credentials_from_role(#aws_config{assume_role = AssumeRoleOptions} = Config) ->
+    RoleArn = proplists:get_value(role_arn, AssumeRoleOptions, undefined),
+    SessionName = proplists:get_value(session_name, AssumeRoleOptions,
+                                      "erlcloud"),
+    SessionExpiration = proplists:get_value(session_expiration,
+                                            AssumeRoleOptions, 900),
+    ExternalId = proplists:get_value(external_id, AssumeRoleOptions,
```

Comment:
We have two options:
1. Add external ID to the cache key {role_metadata, RoleARN, Externald}
2. We keep the external ID cached and if it changes we invalidate the record and get new credentials from sts with the new ExternalID

**Option 1:**
- Cons: Increase the pollution of the set_env cache system if we have many different external_ids
- Optimize the calls of sts:assume role  

**Option 2:**
- Cons: If most of the operations are between two different externalID we will be calling sts:assume_role many times to swap between external ids
- Pro: Less cache pollution

I prefer Option 1